### ISSUE TITLE
radius namespace produces 'empty group' error

### ISSUE THREAD

**Issue by flytomoon** [FreeRADIUS/freeradius-server#1802]
# Issue type
- Questions about the server or its usage should be posted to the [users mailing list](http://freeradius.org/list/users.html).
- Remote security exploits **MUST** be sent to security@freeradius.org.
- [ ] Defect - Crash or memory corruption.
- [ ] Defect - Non compliance with a standards document, or incorrect API usage.
- [x] Defect - Unexpected behaviour (obvious or verified by project member).
- [ ] Feature request.

See [here](https://github.com/FreeRADIUS/freeradius-server/blob/v4.0.x/doc/bugs) for debugging instructions.

# Defect/Feature description
## How to reproduce issue
1. Make the following changes:
In 'default' under server{}
Set namespace = radius

Rename authorize {} to:
	recv Access-Request {}

Authenticate to:
	process Access-Request {}

Post-Auth to:
	send Access-Accept {}
	send Access-Reject {}

 Send a request to the server.

## Output of ``[radiusd|freeradius] -X`` showing issue occurring
(you may need to run ``[radiusd|freeradius] -fxx -l stdout`` if using eg RADIUS with TLS)

```text
Ready to process requests
(1)  Received Access-Request Id 27 from 23.253.111.212:51883 to 174.143.201.142:1812 via eth0 length 247
(1)    Acct-Session-Id = "328"
(1)    NAS-Identifier = "SERVER-04"
(1)    NAS-IP-Address = 23.253.111.212
(1)    Service-Type = Framed-User
(1)    Framed-Protocol = PPP
(1)    NAS-Port = 1
(1)    NAS-Port-Type = Virtual
(1)    Tunnel-Type:0 = L2TP
(1)    Tunnel-Medium-Type:0 = IPv4
(1)    Called-Station-Id = "23.253.111.212"
(1)    Tunnel-Server-Endpoint:0 = "23.253.111.212"
(1)    Calling-Station-Id = "75.82.17.140"
(1)    Tunnel-Client-Endpoint:0 = "75.82.17.140"
(1)    User-Password = "*************"
(1)    MS-Network-Access-Server-Type = Remote-Access-Server
(1)    MS-RAS-Vendor = 311
(1)    MS-RAS-Version = "MSRASV5.20"
(1)    User-Name = "***************"
(1)    Proxy-State = 0xfe800000000000004422ec9f19633c35000001cf
(1)  Running 'recv Access-Request' from file /usr/local/etc/raddb/sites-enabled/default
(1)  empty-group {
(1)    } # empty-group ... <ignoring empty subsection>
(1)  }
(1)  Invalid user: [******************] (from client xxxxxxxxx-MSFT-Server port 1 cli 75.82.17.140)
(1)  Running 'send Access-Reject' from file /usr/local/etc/raddb/sites-enabled/default
(1)  empty-group {
(1)    } # empty-group ... <ignoring empty subsection>
(1)  }
(1)  ERROR: Empty instruction.  Hard-coding to reject.
(1)  Sent Access-Reject Id 27 from 174.143.201.142:1812 to 23.253.111.212:51883 via eth0 length 20
(1)  Delaying response for 1.000000 seconds
(1)  Waiting for 1.000000 seconds
(1)  Failed inserting delay event: Invalid arguments (NULL event list)
(1)  Waiting for 5.000000 seconds
(1)  Failed inserting delay event: Invalid arguments (NULL event list)
(1)  Cleaning up request packet ID 27 with timestamp +13
Ready to process requests
```

**Comment by alandekok** [comment]
At this time, v4.0.x should only be used by people who are sending in code patches to it.

If you need a RADIUS server, run version 3.0.12.

**Comment by arr2036** [comment]
this particular use case requires async

**Comment by arr2036** [comment]
it does appear that the latest changes to the interpreter have broken it

**Comment by alandekok** [comment]
If you want to test v4, go ahead.  But be aware that it's in development.

A sample v4 server is in raddb/sites-available/radius-acct. You CANNOT just make trivial changes to the configuration files, and expect them to work.

All of this (for now) requires an awareness of the internals of v4.  It's mostly undocumented for a reason: it's still in development, and may change.  I do NOT want people to try v4 in production in it's current state, and then be surprised when a newer version behaves differently.

**Comment by arr2036** [comment]
@flytomoon could you post the current state of your ``sites-available/default`` file, and I'll try to reproduce locally.

**Comment by alandekok** [comment]
I've already pushed a message which catches this.  See commits 7e61e5d4, 03d57f71a, and 95f8dcea7697.

**Comment by arr2036** [comment]
In this case @flytomoon needed to poll an external REST interface to check if a user had interacted with another service.  The delay module allowed the request to be yielded between polling the REST interface.  It seemed like a useful/basic test for the async code, so I provided him with instructions on how to try with the v4 interpreter.  To me his config looked correct... are you saying using both section naming schemes with the radius namespace set will cause issues?

**Comment by alandekok** [comment]
Yes.  Changing using `namespace = radius` and `accounting` or `post-proxy` etc. doesn't work.  I reproduced it here, and pushed changes

**Comment by alandekok** [comment]
And, TBH, the new-style `recv` and `send` implementation is only partially done.  There is no documentation.  There are no example configurations for it.  There are no tests.

It is bad practice to recommend that people use features which are undocumented and in development.

If a *developer* wants to fix a bug in v4.0.x, I'm happy to accept patches.  If someone says that an UNDOCUMENTED and UNFINISHED feature doesn't work as expected.  Well... your expectations are wrong.  The feature is *unfinished*, which means that there should be *no* expectations about what it does.

The functionality may change in the future.  The configuration may change.  The macroscopic behavior may change.

It is entirely unproductive to spend time debugging features which aren't finished.

If you're a developer willing to debug code, use v4.0.x.  Everyone else should use a stable version.  And should NOT file bugs against partially-finished features in v4.0.x.

**Comment by arr2036** [comment]
@flytomoon the issue here appears to be that the old configuration sections were left in ``sites-available/default``. @alandekok has helpfully pushed some changes to the code to highlight this as the problem.

If you remove the old section names e.g. ``preacct {}``, ``accounting {}`` etc.. then it should work.

**Comment by flytomoon** [comment]
Thank you both for your attention to this.  I should mention that we are currently using 3.1.x in our production environment.  We do not have any preconceived expectations for development build (and noted unstable) 4.0.x.  However, we do have great interest in seeing this mature to production use for the reasons Arran mentioned above.  I am happy to personally do what I can in helping to trace any issues I come across to make it easier and more expedient for everyone.

Per the above discussion I pulled the most recent code and removed the preacct{} and accounting{} sections.  I did need to add a 'recv Accounting-Request {}' section.  This successfully runs.  It also produces the same error I received before making the namespace change.  Specifically, the Request event list is null by the time it reaches the delay module.  Should I open a new issue for this?

Thank you again for helping to work through this.
